### ISSUE TITLE
howto create a reproducible package?

### ISSUE THREAD

**Issue by katakombi** [megastep/makeself#159]
Howdy,

 could someone please give a tiny example on how to create an archive byte-to-byte reproducible ?
I couldn't make it work (version 2.4.0) even with the `--packaging-date` switch

thx and greetings

**Comment by realtime-neil** [comment]
tldr; I'm aware of no cross-platform way of doing what you require.

If you have reasonably recent GNU tar on Ubuntu Xenial or Bionic, then you can do this:
```
tar_extra='--numeric-owner --owner=0 --group=0 --format=pax --pax-option=exthdr.name=%d/PaxHeaders.0/%f --pax-option=delete=atime --pax-option=delete=ctime --pax-option=delete=mtime --mtime=@0'

(
    cd "$(mktemp -dt archive_dir.XXXXXX)"
    touch foo bar qux
    "${HOME}/code/makeself/makeself.sh" \
        --sha256 \
        --nocomp \
        --packaging-date @0 \
        --target 'the-target' \
        --tar-extra "${tar_extra}" \
        . \
        /tmp/foobarqux.run \
        'My Self-Extracting Archive' \
        ls -lah
    sha256sum /tmp/foobarqux.run
    stat /tmp/foobarqux.run
)

```

Note that, by default, makeself embeds the packaging date as well as a (nondeterministic) target directory, so you need override those if you care about the hash of the entire self-extracting archive. Also note that the makeself default compression is `gzip`, which (without the `--no-name` option), will compress the archive nondeterministically.

An explanation of the options to `tar` would easily fill a blog post, but suffice it to say that PAX header is complicated and various implementations do wildly different things with it.

**Comment by katakombi** [comment]
Thanks for your suggestion, I'm going to test it ASAP. 
As long as a recent gnu tar version is required just at packaging time that would be perfectly fine to me.

Wouldn't a "reproducible packaging" feature be a great enhancement for makeself, where the shasum of the makeself file could ensure not only the integrity but also the byte-to-byte compatibility to a certain version of a released software? I wouldn't really care about an internal packaging date and instead rely on the file naming for a certain packaging date if this means I can reproduce a certain build at any time and any location.

**Comment by realtime-neil** [comment]
tldr; reproducible packaging via `makeself` is impractical and/or impossible.

Longer explanation follows...

Start with the assumption that there is a project goal that supersedes all others; i.e., `makeself` must support a _very_ large population of Unix and Unix-like platforms. This all-important _raison d'être_ is mentioned in the README.md when it first appears in source control:

https://github.com/megastep/makeself/commit/cb1ed4c01818e6dad0146c2560046bed7df5262d#diff-04c6e90faac2675aa89e2176d2eec7d8R24)

Moreover, in the consideration of new features, @megastep has reiterated this aggressively conservative backward-compatibility:

https://github.com/megastep/makeself/issues/68#issuecomment-381766274

And backward-compatibility is good.

The problem is that, relative to the "classic" Unix distributions, reproducible builds are a "modern" concern. Generally speaking, the tools were made to go fast. Deterministic behavior, if it happened at all, happened by accident or with great effort. Old `tar` implementations running on OSs we've never seen don't even have the options to support deterministic binary output. The `tar` archive format has been (in)famously resistant to determinism. The reproducible builds people have an entire page just for `tar`:

https://reproducible-builds.org/docs/archives/

Now, I _love_ me some determinism, which is part of the reason why I wanted a stable order of inputs for `makeself`:

https://github.com/megastep/makeself/pull/153

Furthermore, I want that one-to-one mapping between the archive content and the checksum(s). I don't have to trot out the usual arguments for veracity and security, @katakombi , you've no doubt heard them all before. But, for this tool, for this project, reproducible packaging must be sacrificed for backward-compatibility.

**Comment by katakombi** [comment]
@realtime-neil thanks for the clarification. Judging from what I know about makeself I already had this expected. I just didn't want to give up on it too easily.

**Comment by realtime-neil** [comment]
@katakombi it's a pleasure to discuss things like this with those that care about them. I do hope you can make the aforementioned `tar` trickery work for you. Can this issue be closed?

**Comment by katakombi** [comment]
Thanks a lot! Tested anc confirmed to work!

**Comment by realtime-neil** [comment]
@katakombi sorry for the lateness of my reply. The usual reasons to recommend pax over other formats is that it's standardized. Other reasons might be that, if you can stay under the 100-character filename limit, the "extended headers" are not required --- and most implementations don't generate them. This produces an archive format equivalent to ustar, which is the bog standard of at least the last two decades.
